### DISCUSSION TITLE
MlflowModelSaverDataSet and MlflowArtifactDataset saves to wrong directory after pulling from repo.

### DISCUSSION THREAD

**Discussion by ciaciura** [Galileo-Galilei/kedro-mlflow#456] [Q&A] [answered]
I when i pulled a project from repo and tried to run it threw a "[WinError 5] Access is denied 'path/to/project/on/different/pc' ",  when saving MlflowModelSaverDataSet and MlflowArtifactDataset. Other files saved by kedro in the pipeline prior, generated properly with no errors. Prior to running i recreated conda enviroment from file and ran kedro mlflow init. 

How does kedro-mlflow infer the project directory and can i somehow change it?

Traceback:
`
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\io\core.py:220 in save         │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro_mlflow\io\artifacts\mlflow_art │
│ ifact_dataset.py:78 in _save                                                                     │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\mlflow\tracking\fluent.py:874 in     │
│ log_artifact                                                                                     │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\mlflow\tracking\client.py:1092 in    │
│ log_artifact                                                                                     │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\mlflow\tracking\_tracking_service\cl │
│ ient.py:452 in log_artifact                                                                      │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\mlflow\store\artifact\local_artifact │
│ _repo.py:59 in log_artifacts                                                                     │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\mlflow\utils\file_utils.py:192 in    │
│ mkdir                                                                                            │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\mlflow\utils\file_utils.py:189 in    │
│ mkdir                                                                                            │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:213 in makedirs                              │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\os.py:223 in makedirs                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\<user.on.pervious.pc>'

The above exception was the direct cause of the following exception:

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ C:\Users\User\miniconda3\envs\penguin-env\lib\runpy.py:194 in _run_module_as_main                │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\runpy.py:87 in _run_code                           │
│                                                                                                  │
│ in <module>:7                                                                                    │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\framework\cli\cli.py:211 in    │
│ main                                                                                             │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\click\core.py:1157 in __call__       │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\framework\cli\cli.py:139 in    │
│ main                                                                                             │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\click\core.py:1078 in main           │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\click\core.py:1688 in invoke         │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\click\core.py:1434 in invoke         │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\click\core.py:783 in invoke          │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\framework\cli\project.py:453   │
│ in run                                                                                           │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\framework\session\session.py:4 │
│ 35 in run                                                                                        │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\runner\runner.py:103 in run    │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\runner\sequential_runner.py:70 │
│ in _run                                                                                          │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\runner\runner.py:331 in        │
│ run_node                                                                                         │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\runner\runner.py:446 in        │
│ _run_node_sequential                                                                             │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\io\data_catalog.py:533 in save │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\io\core.py:619 in save         │
│                                                                                                  │
│ C:\Users\User\miniconda3\envs\penguin-env\lib\site-packages\kedro\io\core.py:227 in save         │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
DatasetError: Failed while saving data to data set
MlflowMlflowModelSaverDataSet(filepath=C:\Users\User\penguins\data\06_models\classificator, flavor=mlflow.sklearn,
load_args={}, save_args={}).
[WinError 5] Access is denied: 'C:\\Users\\<user.on.pervious.pc>'
`

catalog.yml for data that fails to save is as follows:

classificator:
  type: kedro_mlflow.io.artifacts.MlflowArtifactDataSet
  data_set:
    type: kedro_mlflow.io.models.MlflowModelSaverDataSet
    flavor: mlflow.sklearn
    filepath: data/06_models/classificator

When i remove these lines kedro runs with no issues, but model is obviously not saved, so it has to be a problem with how model is being saved.

package info:
kedro v0.18.13
kedro_mlflow: v0.11.10

**Comment by Galileo-Galilei** [comment]
Hi @ciaciura , sorry I didn't see this notification. it's better to reach me in the issues rather than the github discussion. 

This bug is extremely weird because I don't see how it can "remember" the previous path, it must be stored somewhere. The error comes from mlflow internal. It seems that mlflow does not manage to copy your model to a local "temp folder" before uploading it or something like that. 

- can you show me your ``mlflow.yml``? 
- What is you mlflow tracking uri? Is it a local ``mlruns`` folder? if yes, can you try to delete your local ``mlruns`` folder and rerun the pipeline?Are parameters correctly stored in an mlflow run apart from the model?  
- Do you still get the error if you change your catalog to remove artifact : 

```yaml
classificator:
  type: kedro_mlflow.io.models.MlflowModelSaverDataSet
  flavor: mlflow.sklearn
  filepath: data/06_models/classificator
```
This should at least store the model locally.

**Reply by Galileo-Galilei** [reply, marked as answer]
Found it! I guess you refer to this repo: https://github.com/ciaciura/penguins. You have push your local ``mlruns`` folder so now mlflow will read the artifacts from the "permanent" location where you stored them. This can be seen here : https://github.com/ciaciura/penguins/blob/main/mlruns/0/meta.yaml. I guess this is the path in your old computer. 

You should not push your ``mlruns`` folder which is only made for local persistence (add it to your ``.gitignore`` file). If you want to share your mlflow runs between different computers, you must set up a remote mlflow server.

**Reply by ciaciura** [reply]
I've tried deleting past runs before with mlflow gc but it omits the default "0" run, that you pointed out still persists. After deleting the "0" folder manually everything runs smoothly, will be including mlruns in .gitignore from now on,  thanks a lot!
